torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -685,33 +682,29 @@ def __init__(
 
         self.n_local_stages = len(stages)
         self.rank = stages[0].group_rank
-
-    def _step_microbatches(
-        self,
-        arg_mbs: Optional[List] = None,
-        kwarg_mbs: Optional[List] = None,
-        target_mbs: Optional[List] = None,
-        losses: Optional[List] = None,
-    ):
-        """
-        Operate on the microbatches for interleaved 1f1b schedule (https://arxiv.org/pdf/2104.04473.pdf).
-
-        Highest rank has a warmup (fwd only) count of [len(stages) - 1] * number of PP ranks
-        and each rank away from highest rank adds 2 warmup steps due to:
-            - one happened before highest rank's warmup started,
-            - one waiting for backward result to trickle down from highest rank
-
-        TODO: Interleaved 1F1B does not support using _sorted_batch_p2p()
-        because it requires recvs and sends from different peers
-        to execute in the same coalesced operation. As a result, this schedule does
-        not support models with skip connections.
-        """
-        arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs, losses)
-
-        # increment warmup_steps by 2 for each hop away
-        warmup_steps = (self.n_local_stages - 1) * self.pp_group_size
-        warmup_steps += 2 * ((self.pp_group_size - 1) - self.rank)
-        warmup_steps = min(warmup_steps, self._n_microbatches * self.n_local_stages)
+        self.group = stages[0].group
+
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: List[
+            List[Optional[Tuple[_ComputationType, int, int]]]
+        ] = []
+        # ========================================================================
+        for rank in range(self.pp_group_size):
+            rank_ops = self._calculate_single_rank_operations(rank)
+            self.pipeline_order.append(rank_ops)
+
+    def _calculate_single_rank_operations(
+        self, rank
+    ) -> List[Optional[Tuple[_ComputationType, int, int]]]:
+        def get_rank_warmup_steps(rank):
+            # Increment warmup_steps by 2 for each hop away
+            warmup_steps = (self.n_local_stages - 1) * self.pp_group_size
+            warmup_steps += 2 * ((self.pp_group_size - 1) - rank)
+            return min(warmup_steps, self._n_microbatches * self.n_local_stages)
+
+        warmup_steps = get_rank_warmup_steps(rank)
         fwd_bwd_steps = (self.n_local_stages * self._n_microbatches) - warmup_steps
         cooldown_steps = (self.n_local_stages * self._n_microbatches) - fwd_bwd_steps
```

Comment:
The PR says "steps" are normalized now, meaning each step can be only 1 forward or 1 backward. Are the `*_steps` here normalized now? Asking because of the `assert` below:
```
        assert (
            warmup_steps + fwd_bwd_steps * 2 + cooldown_steps
            == self.n_local_stages * self._n_microbatches * 2
        )
```
There is a `fwd_bwd_steps * 2`

Reply:
There "steps" are not normalized, so I updated the variable names to "warmup_ops", "1f1b_ops", etc. instead of steps to make it more clear